// forest-express/dist/integrations/index.js
"use strict";

var CloseIo = require('./close.io');
var Intercom = require('./intercom');
var Layer = require('./layer');
var MixPanel = require('./mixpanel');
var Stripe = require('./stripe');
function IntegrationChecker(opts, Implementation) {
  var modules = [new CloseIo(opts, Implementation), new Intercom(opts, Implementation), new Layer(opts, Implementation), new MixPanel(opts, Implementation), new Stripe(opts, Implementation)];
  this.defineRoutes = function (app, model) {
    modules.forEach(function (module) {
      if (module.defineRoutes) {
        module.defineRoutes(app, model);
      }
    });
  };
  this.defineCollections = function (collections) {
    modules.forEach(function (module) {
      if (module.defineCollections) {
        module.defineCollections(collections);
      }
    });
  };
  this.defineSegments = function (model, schema) {
    modules.forEach(function (module) {
      if (module.defineSegments) {
        module.defineSegments(model, schema);
      }
    });
  };
  this.defineFields = function (model, schema) {
    modules.forEach(function (module) {
      if (module.defineFields) {
        module.defineFields(model, schema);
      }
    });
  };
  this.defineSerializationOption = function (model, schema, dest, field) {
    modules.forEach(function (module) {
      if (module.defineSerializationOption) {
        module.defineSerializationOption(model, schema, dest, field);
      }
    });
  };
}
module.exports = IntegrationChecker;
