// @forestadmin/datasource-toolkit/dist/src/interfaces/query/filter/factory.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const luxon_1 = require("luxon");
const unpaginated_1 = __importDefault(require("./unpaginated"));
const collection_1 = __importDefault(require("../../../utils/collection"));
const schema_1 = __importDefault(require("../../../utils/schema"));
const factory_1 = __importDefault(require("../condition-tree/factory"));
const leaf_1 = __importDefault(require("../condition-tree/nodes/leaf"));
const projection_1 = __importDefault(require("../projection"));
class FilterFactory {
    static getPreviousConditionTree(field, startPeriod, endPeriod) {
        return factory_1.default.intersect(new leaf_1.default(field, 'GreaterThan', startPeriod.toISO()), new leaf_1.default(field, 'LessThan', endPeriod.toISO()));
    }
    static getPreviousPeriodByUnit(field, now, interval) {
        const dayBeforeYesterday = now.minus({ [interval]: 2 });
        return this.getPreviousConditionTree(field, dayBeforeYesterday.startOf(interval), dayBeforeYesterday.endOf(interval));
    }
    static getPreviousPeriodFilter(filter, timezone) {
        const now = luxon_1.DateTime.now().setZone(timezone);
        return filter.override({
            conditionTree: filter.conditionTree.replaceLeafs(leaf => {
                switch (leaf.operator) {
                    case 'Today':
                        return leaf.override({ operator: 'Yesterday' });
                    case 'Yesterday':
                        return this.getPreviousPeriodByUnit(leaf.field, now, 'day');
                    case 'PreviousWeek':
                        return this.getPreviousPeriodByUnit(leaf.field, now, 'week');
                    case 'PreviousMonth':
                        return this.getPreviousPeriodByUnit(leaf.field, now, 'month');
                    case 'PreviousQuarter':
                        return this.getPreviousPeriodByUnit(leaf.field, now, 'quarter');
                    case 'PreviousYear':
                        return this.getPreviousPeriodByUnit(leaf.field, now, 'year');
                    case 'PreviousXDays': {
                        const startPeriodXDays = now.minus({ days: 2 * Number(leaf.value) });
                        const endPeriodXDays = now.minus({ days: Number(leaf.value) });
                        return this.getPreviousConditionTree(leaf.field, startPeriodXDays.startOf('day'), endPeriodXDays.startOf('day'));
                    }
                    case 'PreviousXDaysToDate': {
                        const startPeriod = now.minus({ days: 2 * Number(leaf.value) });
                        const endPeriod = now.minus({ days: Number(leaf.value) });
                        return this.getPreviousConditionTree(leaf.field, startPeriod.startOf('day'), endPeriod);
                    }
                    case 'PreviousWeekToDate':
                        return leaf.override({ operator: 'PreviousWeek' });
                    case 'PreviousMonthToDate':
                        return leaf.override({ operator: 'PreviousMonth' });
                    case 'PreviousQuarterToDate':
                        return leaf.override({ operator: 'PreviousQuarter' });
                    case 'PreviousYearToDate':
                        return leaf.override({ operator: 'PreviousYear' });
                    default:
                        return leaf;
                }
            }),
        });
    }
    /**
     * Make a filter targeting the through collection of a many to many relationship from the relation
     * and a filter to the target collection.
     */
    static async makeThroughFilter(collection, id, relationName, caller, baseForeignFilter) {
        const relation = collection.schema.fields[relationName];
        const originValue = await collection_1.default.getValue(collection, caller, id, relation.originKeyTarget);
        const foreignRelation = collection_1.default.getThroughTarget(collection, relationName);
        // Optimization for many to many when there is not search/segment (saves one query)
        if (foreignRelation && baseForeignFilter.isNestable) {
            const baseThroughFilter = baseForeignFilter.nest(foreignRelation);
            return baseThroughFilter.override({
                conditionTree: factory_1.default.intersect(new leaf_1.default(relation.originKey, 'Equal', originValue), new leaf_1.default(relation.foreignKey, 'Present'), baseThroughFilter.conditionTree),
            });
        }
        // Otherwise we have no choice but to call the target collection so that search and segment
        // are correctly apply, and then match ids in the though collection.
        const target = collection.dataSource.getCollection(relation.foreignCollection);
        const records = await target.list(caller, await FilterFactory.makeForeignFilter(collection, id, relationName, caller, baseForeignFilter), new projection_1.default(relation.foreignKeyTarget));
        return new unpaginated_1.default({
            conditionTree: factory_1.default.intersect(
            // only children of parent
            new leaf_1.default(relation.originKey, 'Equal', originValue), 
            // only the children which match the conditions in baseForeignFilter
            new leaf_1.default(relation.foreignKey, 'In', records.map(r => r[relation.foreignKeyTarget]))),
        });
    }
    /**
     * Given a collection and a OneToMany/ManyToMany relation, generate a filter which
     * - match only children of the provided recordId
     * - can apply on the target collection of the relation
     */
    static async makeForeignFilter(collection, id, relationName, caller, baseForeignFilter) {
        const relation = schema_1.default.getToManyRelation(collection.schema, relationName);
        const originValue = await collection_1.default.getValue(collection, caller, id, relation.originKeyTarget);
        // Compute condition tree to match parent record.
        let originTree;
        if (relation.type === 'OneToMany') {
            // OneToMany case (can be done in one request all the time)
            originTree = new leaf_1.default(relation.originKey, 'Equal', originValue);
        }
        else {
            // ManyToMany case (more complicated...)
            const through = collection.dataSource.getCollection(relation.throughCollection);
            const throughTree = factory_1.default.intersect(new leaf_1.default(relation.originKey, 'Equal', originValue), new leaf_1.default(relation.foreignKey, 'Present'));
            const records = await through.list(caller, new unpaginated_1.default({ conditionTree: throughTree }), new projection_1.default(relation.foreignKey));
            originTree = new leaf_1.default(relation.foreignKeyTarget, 'In', records.map(r => r[relation.foreignKey]));
        }
        // Merge with existing filter.
        return baseForeignFilter.override({
            conditionTree: factory_1.default.intersect(baseForeignFilter.conditionTree, originTree),
        });
    }
}
exports.default = FilterFactory;
